app/uk/gov/hmrc/hec/services/scheduleService/HecTaxCheckExtractionService.scala

```diff
@@ -42,57 +44,123 @@ class HecTaxCheckExtractionServiceImpl @Inject() (
+            LicenceTimeTrading,
+            seqNum,
+            licenceTimeTrading.partialFileName,
+            licenceTimeTrading.dirName,
+            false
+          )
         _                <- createAndStoreFile(
                               LicenceValidityPeriod,
                               seqNum,
                               licenceValidityPeriod.partialFileName,
-                              licenceValidityPeriod.dirName
+                              licenceValidityPeriod.dirName,
+                              false
                             )
-        //updating isExtracted to true for the the processed hec tax check codes
-        newHecTaxCheck   <- taxCheckService.updateAllHecTaxCheck(updatedHecTaxChek)
-      } yield newHecTaxCheck
+        _                <- createAndStoreFile(
+                              CorrectiveAction,
+                              seqNum,
+                              correctiveAction.partialFileName,
+                              correctiveAction.dirName,
+                              false
+                            )
+        hecTaxCheck      <- taxCheckService.getAllTaxCheckCodesByExtractedStatus(false)
+        _                <- createHecFile(HECTaxCheckFileBodyList(hecTaxCheck), count, hecData.partialFileName, hecData.dirName)
+        updatedHecTaxChek = hecTaxCheck.map(_.copy(isExtracted = true))
+
+        //i think we don't need to bring the whole list of codes which are updated ,
+        // it may cause performance issue, so now it's returning Unit.
+        _ <- taxCheckService.updateAllHecTaxCheck(updatedHecTaxChek)
+      } yield ()
     }
     result.value
   }
 
+  private def toFormattedString(i: Int) = f"$i%04d"
+
+  private def createHecFile(
+    hecTaxCheckList: HECTaxCheckFileBodyList,
+    count: Int,
+    partialFileName: String,
+    dirname: String
+  ): EitherT[Future, Error, List[Unit]] = {
+    @tailrec
+    def loopHecTaxCheckRecords(
+      hecTaxCheckList: List[HECTaxCheck],
+      count: Int,
+      seqNumInt: Int,
+      partialFileName: String,
+      dirname: String
+    ): List[EitherT[Future, Error, Unit]] = {
+      val hecList       = hecTaxCheckList.take(count)
+      val remainingList = hecTaxCheckList.drop(count)
+      val isRemaining   = remainingList.size > 0
+      //If list is empty or sequence number exceeds 9999, then halt the process
+      if (hecList.size === 0 || seqNumInt > 9999) {
+        logger.info(
+          "Hec tax Check file creation process halted:: Either there are no more records to process or the sequence number exceeds 9999. "
+        )
+        List(EitherT.fromEither[Future](Right(())))
+      } else {
+        val _ = createAndStoreFile(
+          HECTaxCheckFileBodyList(hecList),
+          toFormattedString(seqNumInt),
+          partialFileName,
+          dirname,
+          isRemaining
+        )
+        loopHecTaxCheckRecords(remainingList, count, seqNumInt + 1, partialFileName, dirname)
```

Comment:
in order to build the `List` here properly so that the `createAndStoreFile` `EitherT` is not thrown away you could do something to the effect of:
```scala
def loopHecTaxCheckRecords(
      hecTaxCheckList: List[HECTaxCheck],
      count: Int,
      seqNumInt: Int,
      partialFileName: String,
      dirname: String,
      acc: List[EitherT[Future, Error, Unit]]
    ): List[EitherT[Future, Error, Unit]] = {
 if(isEnd)
    acc
else
    loopHecTaxCheckRecords(..., createAndStoreFile(...) :: acc)
}
```